Factor out gradient code out into strongly typed CSS/Style values

weinig opened the pull request:
#### b95f33ec7bcec3053fe9f27e1872644e8db12bfd
<pre>
Factor out gradient code out into strongly typed CSS/Style values
<a href="https://bugs.webkit.org/show_bug.cgi?id=280901">https://bugs.webkit.org/show_bug.cgi?id=280901</a>

Reviewed by Darin Adler.

Factors CSS and Style gradient code into their own top level types,
retaining the existing CSSGradient*Value and StyleGradientImage types
as thin wrappers.

Transitions overload based customization points use for CSS/Style type
operations to instead use template specialization based customization
points. This allows for either to understand error messages and does not
require ADL wizardry to get right. It also allows us to separate the
signature of the customization point from the signature of the call that
invokes the customization point, making it easier for default values.

To reduce duplicate code, default implementations of some of the customization
points (VisitCSSValueChildren, ComputedStyleDependenciesCollector, conversions)
have default implementations for &quot;tuple-like&quot; types (those that implement the
&quot;tuple-like&quot; pseudo protocol). For the conversions, an additional step of
defining the mapping between the types is needed and a helper macro is provided.

* Source/WTF/wtf/StdLibExtras.h:
    Add concept for &quot;tuple-like&quot;.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add files.
* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/color/CSSAbsoluteColorResolver.h:
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/values/CSSGradient.cpp: Added.
* Source/WebCore/css/values/CSSGradient.h: Added.
* Source/WebCore/css/values/CSSPosition.cpp:
* Source/WebCore/css/values/CSSPosition.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+CSSValueVisitation.h: Added.
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/values/StyleGradient.cpp: Added.
* Source/WebCore/style/values/StyleGradient.h: Added.
* Source/WebCore/style/values/StylePosition.cpp:
* Source/WebCore/style/values/StylePosition.h:
* Source/WebCore/style/values/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/StyleValueTypes.h:

Canonical link: <a href="https://commits.webkit.org/285038@main">https://commits.webkit.org/285038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/caa05d8730cdb4c13a7f4b193aed96428e347443

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71198 "136 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23971 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75305 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22403 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73313 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56292 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14749 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74264 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45984 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61361 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36727 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42657 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18823 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20743 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64323 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19186 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77025 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70447 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15432 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18367 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64001 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15474 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63977 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15782 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12122 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5751 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92227 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46411 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1190 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20107 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47482 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSGradientValue.cpp:
```diff
@@ -26,1029 +26,248 @@
-    result.append(interleave(m_stops, appendColorStop, ", "_s), ')');
-
-    return result.toString();
+    StringBuilder builder;
+    CSS::serializationForCSS(builder, m_gradient);
+    return builder.toString();
```
After seeing this idiom 3 times in quick succession I think maybe we want a function template that does this. Or alternatively a way to write `makeString(CSS::serializationForCSS(m_gradient))`.

darinadler commented on Source/WebCore/css/values/CSSGradient.h:
```diff
@@ -0,0 +1,419 @@
+
+using DeprecatedGradientPosition = SpaceSeparatedTuple<PercentageOrNumber, PercentageOrNumber>;
+
+using Horizontal = std::variant<Left, Right>;
+using Vertical   = std::variant<Top, Bottom>;
+
+using ClosestCorner  = Constant<CSSValueClosestCorner>;
+using ClosestSide    = Constant<CSSValueClosestSide>;
+using FarthestCorner = Constant<CSSValueFarthestCorner>;
+using FarthestSide   = Constant<CSSValueFarthestSide>;
+using Contain        = Constant<CSSValueContain>;
+using Cover          = Constant<CSSValueCover>;
+
+using RadialGradientExtent         = std::variant<ClosestCorner, ClosestSide, FarthestCorner, FarthestSide>;
+using PrefixedRadialGradientExtent = std::variant<ClosestCorner, ClosestSide, FarthestCorner, FarthestSide, Contain, Cover>;
```
I never line up things like this. Maybe because I hope to some day use non-monospace font to edit my code. Apparently other people still like lining things up.

weinig replied:
I do a lot of multiline selection (using option to get that good rectangular selection), so lining things up like this helps as I edit.

darinadler commented on Source/WebCore/css/values/CSSPrimitiveNumericTypes+CSSValueVisitation.h:
```diff
@@ -0,0 +1,49 @@
+};
+
+template<RawNumeric RawType> struct CSSValueChildrenVisitor<PrimitiveNumeric<RawType>> {
+    IterationStatus operator()(const Function<IterationStatus(CSSValue&)>& func, const PrimitiveNumeric<RawType>& value)
```
Maybe use an entire word, `function`?

darinadler commented on Source/WebCore/css/values/CSSUnevaluatedCalc.h:
```diff
@@ -155,24 +156,30 @@ template<typename T> decltype(auto) simplifyUnevaluatedCalc(const std::optional<
-    return func(calc.calc.get());
-}
+template<typename T> struct CSSValueChildrenVisitor<UnevaluatedCalc<T>> {
+    inline IterationStatus operator()(const Function<IterationStatus(CSSValue&)>& func, const UnevaluatedCalc<T>& value)
```
Same thought about the word `function` as above, but maybe `func` is enough a term of art?

weinig replied:
Yeah, pretty much every single one of these visitors uses it.

darinadler commented on Source/WebCore/css/values/CSSValueTypes.h:
```diff
@@ -33,8 +33,86 @@
+// Types can specialize this and set the value to true to be treated as "tuple-like"
+// for CSS value type algorithms.
+// NOTE: This gets automatically specialized when using the CSS_TUPLE_LIKE_CONFORMANCE macro.
+template<class> inline constexpr bool TreatAsTupleLike = false;
```
Does the `inline` here have an effect?

weinig replied:
Yeah. I recently learned this. For global constexpr variables, `inline` tells the linker it should merge instances. (see https://abseil.io/tips/168 or https://higaski.at/constexpr-does-not-imply-inline/). In practice, I think this only matters if we ever take the address of this and want it to be equal, but it seems like its worth the couple letters, and seems like a pretty general rule across other projects that constexpr globals in headers should be inline. (In a fun turn, if the constexpr is a static member variable of a class/struct, inline is implied!)

darinadler commented on Source/WebCore/style/values/StyleValueTypes.h:
```diff
@@ -39,6 +40,121 @@ namespace Style {
+// Types can specialize this and set the value to true to be treated as "non-converting"
+// for css to style / style to css conversion algorithms. This means the type is identical
+// for both CSS and Style systems (e.g. a constant value or an enum).
+template<class> inline constexpr bool TreatAsNonConverting = false;
```
`class` -> `typename`

Does `inline` do something here?

darinadler commented on Source/WebCore/style/values/StyleValueTypes.h:
```diff
@@ -39,6 +40,121 @@ namespace Style {
+// Types can specialize this and set the value to true to be treated as "tuple-like"
+// for css to style / style to css conversion algorithms.
+// NOTE: This gets automatically specialized when using the STYLE_TUPLE_LIKE_CONFORMANCE macro.
+template<class> inline constexpr bool TreatAsTupleLike = false;
```
`class` -> `typename`

Does `inline` do something here?

weinig committed:
Factor out gradient code out into strongly typed CSS/Style values
https://bugs.webkit.org/show_bug.cgi?id=280901

Reviewed by Darin Adler.

Factors CSS and Style gradient code into their own top level types,
retaining the existing CSSGradient*Value and StyleGradientImage types
as thin wrappers.

Transitions overload based customization points use for CSS/Style type
operations to instead use template specialization based customization
points. This allows for either to understand error messages and does not
require ADL wizardry to get right. It also allows us to separate the
signature of the customization point from the signature of the call that
invokes the customization point, making it easier for default values.

To reduce duplicate code, default implementations of some of the customization
points (VisitCSSValueChildren, ComputedStyleDependenciesCollector, conversions)
have default implementations for "tuple-like" types (those that implement the
"tuple-like" pseudo protocol). For the conversions, an additional step of
defining the mapping between the types is needed and a helper macro is provided.

* Source/WTF/wtf/StdLibExtras.h:
    Add concept for "tuple-like".

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add files.
* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/color/CSSAbsoluteColorResolver.h:
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/values/CSSGradient.cpp: Added.
* Source/WebCore/css/values/CSSGradient.h: Added.
* Source/WebCore/css/values/CSSPosition.cpp:
* Source/WebCore/css/values/CSSPosition.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+CSSValueVisitation.h: Added.
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/values/StyleGradient.cpp: Added.
* Source/WebCore/style/values/StyleGradient.h: Added.
* Source/WebCore/style/values/StylePosition.cpp:
* Source/WebCore/style/values/StylePosition.h:
* Source/WebCore/style/values/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/StyleValueTypes.h:

Canonical link: https://commits.webkit.org/285038@main
